### PR TITLE
Add Configuration Annotations Page

### PR DESCRIPTION
**Pull request by RlonRyan**
This commit adds a basic description of the annotation-based mod configuration system that was added in the 1.12.x versions of Minecraft Forge.

Closes MinecraftForge/Documentation#119.

### CODE REVIEW

File: docs/utilities/config_annotations.md [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+@Config(modid = "my_mod_id")
+public class MyModConfig {
+
+  @Comment("This property belongs to the default group.")
+  public static boolean myPropertyDefaultGroup = false;
+
+  public static class MyCategory {
+
+    @Comment("This property belongs to the MyCategory group.")
+    public static boolean myPropertyMyCategory = false;
```
Comment (TehNut) [inline docs/utilities/config_annotations.md]: Values within categories should not be static.

File: docs/utilities/config_annotations.md [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+
+
+Basics
+-------------------------
+
+The annotation-based configuration system revolves around the `@Config` annotation, found in the `net.minecraftforge.common.config` package of Minecraft Forge. The `@Config` annotation is intended to be applied to any class that contains user-configurable fields, requiring only the id of the mod that owns the configuration for basic usage. Following the application of the `@Config` annotation to a configuration class, Forge will then resolve and manage any `public static` fields in the class as configuration elements.
+
+!!! important
+
+    There is currently no way to have Forge ignore a `public static` field in a `@Config` class, so make sure not to place any non-user configurable `public static`
```
Comment (tterrag1098) [inline docs/utilities/config_annotations.md]: incomplete sentence?

File: docs/utilities/config_annotations.md [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+````Java
+@Config(modid = "my_mod_id")
+public class MyModConfig {
+
+  public static boolean myBooleanSetting = false;
+
+}
+````
+
+The prior usage would then create a boolean property `general.myBooleanSetting` with a default value of `false` in a configuration file named `my_mod_id.conf` which is managed by Forge.
```
Comment (tterrag1098) [inline docs/utilities/config_annotations.md]: `.conf` -> `.cfg`

File: docs/utilities/config_annotations.md [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+
+  @Comment({"Whereas this,", "is a multi-line comment"})
+  public static boolean myPropertyWithMultiLineComment = false;
+
+}
+````
+
+!!! note
+
+    Single-line comments, having only one string value set in the annotation, do not require (but will accept) surrounding brackets unlike multi-line comments that require surrounding brackets.
```
Comment (tterrag1098) [inline docs/utilities/config_annotations.md]: "brackets" -> "braces"

File: docs/utilities/config_annotations.md [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+!!! important
+
+    It is strongly recommended that you do not use the newline character `\n` (or platform-specific equivalents) in comments, and instead use the multi-line `@Comment` annotation form, as Forge will then be able to ensure the proper handling of the multi-line comment.
+
+Property Ranges
+-------------------------
+
+While when using boolean configuration properties all possible input values are considered valid, in the case of numeric inputs there may arise the need as to constrain the acceptable inputs to a certain continuous range. For this reason, the annotations `@RangeInt` and `@RangeDouble` exist.
+
+The `@Range` annotations consist of two values, a minimum value and a maximum value, which together define an inclusive range of acceptable inputs for a given numeric property. Note, that inclusive means that the values set for `min` and `max` are considered valid inputs for the property. If either min or max are not set, then the annotation will default to `<? extends Number>.MIN_VALUE` and `<? extends Number>.MAX_VALUE` respectively.
```
Comment (tterrag1098) [inline docs/utilities/config_annotations.md]: I think the generic syntax here is confusing and unnecessary, just say "that data type's min and max values" or something to that effect.

File: docs/utilities/config_annotations.md [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+public class MyModConfig {
+
+  @Comment("This integer property accepts values 1 to 10 inclusive.")
+  @RangeInt(min = 1, max = 10)
+  public static int myIntegerProperty = false;
+
+  @Comment("This double property accepts values 0.0 to 1.0 inclusive.")
+  @RangeDouble(min = 0.0, max = 1.0)
+  public static double myDoubleProperty = false;
+
```
Comment (tterrag1098) [inline docs/utilities/config_annotations.md]: I think these code examples are getting a bit superfluous at this point, try to keep the docs more focused on documenting, not examples.

File: docs/utilities/config_annotations.md [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+}
+````
+
+!! important
+
+    While both the `@RangeInt` and `@RangeDouble` accept a min and a max value to be set they are not interchangeable, given that the type must match. Consequently the `@RangeInt` annotation is only to be applied to `int` and `Integer` fields, and the `@RangeDouble` is only to be applied to `double` and `Double` fields.
+
+!! important
+
+    There currently exists no range annotations capable of supporting numeric types other than the aforementioned `int`, `Integer`, `double`, and `Double` types. As such, you are forced to use `int` instead of `byte` or `long`, `Integer` instead of `Byte` or `Long`, `double` instead of `float`, and `Double` instead of `Float`. This issue will likely be fixed in a future release given wide enough adoption of the new annotation-based configuration system.
```
Comment (tterrag1098) [inline docs/utilities/config_annotations.md]: Source?

File: docs/utilities/config_annotations.md [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+  @Comment("This property would normally have a horrible name!")
+  @Name("my_well_named_property")
+  public static boolean MY_hoRibbliyNAMEdVairable = false;
+
+}
+````
+
+!! Note
+
+    If you are following conventional Java variable naming practices then the variables defined in your `@Config` class likely contain a great number of capital letters, as seen in the provided example usages (prior to the horribly named one). Since the Forge standard for registration names is all-lowercase, the default property names would therefore be unsuitable, and use of the `@Name` annotation is highly recommended.
```
Comment (tterrag1098) [inline docs/utilities/config_annotations.md]: What do registry names have to do with config keys?

File: docs/utilities/config_annotations.md [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+
+    This behavior is partly an oversight and partly a technical limitation. Forge has no intention of punishing any modder who conforms to standard practices. Ideally future releases of Forge will automatically convert traditionally-named variables into conforming property names by replacing capital characters with respective lowercase characters preceded by the underscore character `_`.
+
+!! Note
+
+    The `@Name` annotation cannot currently be applied to a inner class grouping, an oversight which is likely to change in a future version of Forge.
```
Comment (tterrag1098) [inline docs/utilities/config_annotations.md]: I think the docs should avoid speculating on future inclusions, just state the facts as they are currently.

File: docs/utilities/config_annotations.md [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+  @Comment("This property is always changeable!")
+  public static boolean myAlwaysChangeableProperty = false;
+
+  @Comment("This property is only changeable on world reloads!")
+  @RequiresWorldRestart
+  public static boolean myWorldReloadChangeableProperty = false;
+
+  @Comment("This property is only changeable on Minecraft restarts!")
+  @RequiresMcRestart
+  public static boolean myMcReloadChangeableProperty = false;
```
Comment (tterrag1098) [inline docs/utilities/config_annotations.md]: This code snippet does not introduce any new information. A reader knows what an annotation is, how it is used, and what the phrases "world reload" and "minecraft restart" mean.

File: docs/utilities/config_annotations.md [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+}
+````
+
+!! Important
+
+    Keep in mind that for installs including more than just one or two other mods can take a great deal of time to restart. As such you should try to make as many properties as possible changeable on the fly.
+
+!! Important
+
+    While this annotation can be applied to categories, it is strongly recommended that you do not, given that the overall goal should be to minimize the frequency of reloads, which take a long time.
```
Comment (tterrag1098) [inline docs/utilities/config_annotations.md]: Superfluous comment, the above covers this. If it needs to be mentioned, just say above that it can be used on categories.

**Comment by PaleoCrafter** [comment]
As a general comment regarding these "probably going to be addressed in a future version of Forge": Create an issue or - even better - a PR that actually adresses these rather than pointing it out in the docs :P

File: docs/utilities/config_annotations.md (hunk 7 above)
Comment (TehNut) [reply docs/utilities/config_annotations.md]: Yeah, don't speculate in documentation. We can always add them later.

File: docs/utilities/config_annotations.md [hunk 12]
```diff
@@ -0,0 +1,234 @@
+Configuration Annotations
+=========================
+
+The 1.12.x version of Minecraft Forge introduced novel mod configuration system based on annotations. This new system relies on Forge ASM as to automatically discover mod configuration classes and their respective configurable fields as to make working with configurations as easy as possible. Ideally this new system will encourage the standard implementation of mod configurations as to afford end-users with maximally configurable modded Minecraft experiences.
```
Comment (elifoster) [inline docs/utilities/config_annotations.md]: This last sentence is weird. We've had configs forever, and either way speculation isn't good to have in docs.

File: docs/utilities/config_annotations.md [hunk 13]
```diff
@@ -0,0 +1,234 @@
...
+
+
+Basics
+-------------------------
+
+The annotation-based configuration system revolves around the `@Config` annotation, found in the `net.minecraftforge.common.config` package of Minecraft Forge. The `@Config` annotation is intended to be applied to any class that contains user-configurable fields, requiring only the id of the mod that owns the configuration for basic usage. Following the application of the `@Config` annotation to a configuration class, Forge will then resolve and manage any `public static` fields in the class as configuration elements.
```
Comment (elifoster) [inline docs/utilities/config_annotations.md]: >of Minecraft Forge

in the first sentence is unnecessary

Comment (elifoster) [inline docs/utilities/config_annotations.md]: Regarding `public static`, does it actually have to be public?

File: docs/utilities/config_annotations.md (hunk 2 above)
Comment (elifoster) [inline docs/utilities/config_annotations.md]: Regarding `public static`, does it actually have to be public?

File: docs/utilities/config_annotations.md [hunk 14]
```diff
@@ -0,0 +1,234 @@
...
+Consequently, the simplest possible usage of the annotation-based configuration system would be as follows.
+
+````Java
+@Config(modid = "my_mod_id")
+public class MyModConfig {
+
+  public static boolean myBooleanSetting = false;
+
+}
+````
```
Comment (elifoster) [inline docs/utilities/config_annotations.md]: This should only be 3 backticks

File: docs/utilities/config_annotations.md [hunk 15]
```diff
@@ -0,0 +1,234 @@
...
+
+The annotation-based configuration system revolves around the `@Config` annotation, found in the `net.minecraftforge.common.config` package of Minecraft Forge. The `@Config` annotation is intended to be applied to any class that contains user-configurable fields, requiring only the id of the mod that owns the configuration for basic usage. Following the application of the `@Config` annotation to a configuration class, Forge will then resolve and manage any `public static` fields in the class as configuration elements.
+
+!!! important
+
+    There is currently no way to have Forge ignore a `public static` field in a `@Config` class, so make sure not to place any non-user configurable `public static`
+
+Consequently, the simplest possible usage of the annotation-based configuration system would be as follows.
+
+````Java
```
Comment (elifoster) [inline docs/utilities/config_annotations.md]: This should only be 3 backticks

File: docs/utilities/config_annotations.md [hunk 16]
```diff
@@ -0,0 +1,234 @@
...
+@Config(modid = "my_mod_id")
+public class MyModConfig {
+
+  public static boolean myBooleanSetting = false;
+
+}
```
Comment (elifoster) [inline docs/utilities/config_annotations.md]: Since this is just a brief example, the two empty lines are probably unnecessary and take up room.

File: docs/utilities/config_annotations.md [hunk 17]
```diff
@@ -0,0 +1,234 @@
...
+````
+
+The prior usage would then create a boolean property `general.myBooleanSetting` with a default value of `false` in a configuration file named `my_mod_id.conf` which is managed by Forge.
+
+Whereas the prior example used only the Java primitive `boolean`, the annotation-based configuration system will accept all Java primitives, and their respective boxed forms. Furthermore, the annotation system is additionally capable of accepting `String` and `enum` values.
```
Comment (elifoster) [inline docs/utilities/config_annotations.md]: In the last sentence, "additionally" is redundant.

File: docs/utilities/config_annotations.md [hunk 18]
```diff
@@ -0,0 +1,234 @@
...
+
+!! important
+
+    By default the annotation-based configuration system loads all mod configurations exactly once, directly after the mod is constructed but before the mod enters pre-initialization.
+
+
+Comments
+-------------------------
+
+Often, despite our best efforts in naming, the property key alone does not do justice in describing the function of a given configuration property. In this case, or moreso in general it would be nice as to be able to give the end-user more information as to what a configuration element actually does. For this reason, the `@Comment` annotation was created.
```
Comment (elifoster) [inline docs/utilities/config_annotations.md]: This has a lot of unnecessary fluff. I would reword it as

>The `@Comment` annotation can be used to add additional information to configuration properties that might not fit in the property key alone.

File: docs/utilities/config_annotations.md [hunk 19]
```diff
@@ -0,0 +1,234 @@
...
+
+The `@Comment` annotation is simple to use, as it may be applied directly to any valid field inside a `@Config` class as to add either a single line, or a multi-line comment. Extending the prior example usage of the `@Config` annotation:
```
Comment (elifoster) [inline docs/utilities/config_annotations.md]: This also has a lot of unnecessary fluff. I would reword it as

>The `@Comment` annotation can be applied to any field inside a class annotated with `@Config`. A `String` can be provided as its argument for a single line comment, or an array of `String`s can be provided as its argument for a multi-line comment where each value in the array is one line.

File: docs/utilities/config_annotations.md [hunk 20]
```diff
@@ -0,0 +1,234 @@
...
+
+````Java
```
Comment (elifoster) [inline docs/utilities/config_annotations.md]: This should only be 3 backticks

File: docs/utilities/config_annotations.md [hunk 21]
```diff
@@ -0,0 +1,234 @@
...
+public class MyModConfig {
+
+  @Comment("This is a lovely single-line comment.")
+  public static boolean myPropertyWithSingleLineComment = false;
+
+  @Comment({"Whereas this,", "is a multi-line comment"})
+  public static boolean myPropertyWithMultiLineComment = false;
+
+}
+````
```
Comment (elifoster) [inline docs/utilities/config_annotations.md]: This should only be 3 backticks

File: docs/utilities/config_annotations.md [hunk 22]
```diff
@@ -0,0 +1,234 @@
...
+-------------------------
+
+Often, despite our best efforts in naming, the property key alone does not do justice in describing the function of a given configuration property. In this case, or moreso in general it would be nice as to be able to give the end-user more information as to what a configuration element actually does. For this reason, the `@Comment` annotation was created.
+
+The `@Comment` annotation is simple to use, as it may be applied directly to any valid field inside a `@Config` class as to add either a single line, or a multi-line comment. Extending the prior example usage of the `@Config` annotation:
+
+````Java
+@Config(modid = "my_mod_id")
+public class MyModConfig {
+
```
Comment (elifoster) [inline docs/utilities/config_annotations.md]: This blank line is probably unnecessary and takes up room

File: docs/utilities/config_annotations.md [hunk 23]
```diff
@@ -0,0 +1,234 @@
...
+  @Comment("This is a lovely single-line comment.")
+  public static boolean myPropertyWithSingleLineComment = false;
+
+  @Comment({"Whereas this,", "is a multi-line comment"})
+  public static boolean myPropertyWithMultiLineComment = false;
+
```
Comment (elifoster) [inline docs/utilities/config_annotations.md]: This blank line is probably unnecessary and takes up room

File: docs/utilities/config_annotations.md [hunk 24]
```diff
@@ -0,0 +1,234 @@
...
+    Single-line comments, having only one string value set in the annotation, do not require (but will accept) surrounding brackets unlike multi-line comments that require surrounding brackets.
+
+!!! important
+
+    It is strongly recommended that you do not use the newline character `\n` (or platform-specific equivalents) in comments, and instead use the multi-line `@Comment` annotation form, as Forge will then be able to ensure the proper handling of the multi-line comment.
+
+Property Ranges
+-------------------------
+
+While when using boolean configuration properties all possible input values are considered valid, in the case of numeric inputs there may arise the need as to constrain the acceptable inputs to a certain continuous range. For this reason, the annotations `@RangeInt` and `@RangeDouble` exist.
```
Comment (elifoster) [inline docs/utilities/config_annotations.md]: This first sentence is really confusing and wordy.

File: docs/utilities/config_annotations.md (hunk 5 above)
Comment (elifoster) [inline docs/utilities/config_annotations.md]: I think explaining what an inclusive range is is unnecessary and out of scope for Forge. It's basic math terminology.

File: docs/utilities/config_annotations.md (hunk 6 above)
Comment (elifoster) [reply docs/utilities/config_annotations.md]: Or at the least remove the unnecessary bits like the class definition and `@Comment`

File: docs/utilities/config_annotations.md [hunk 25]
```diff
@@ -0,0 +1,234 @@
...
+
+The `@Range` annotations consist of two values, a minimum value and a maximum value, which together define an inclusive range of acceptable inputs for a given numeric property. Note, that inclusive means that the values set for `min` and `max` are considered valid inputs for the property. If either min or max are not set, then the annotation will default to `<? extends Number>.MIN_VALUE` and `<? extends Number>.MAX_VALUE` respectively.
+
+Building off the prior examples a sample usage of the `@RangeInt` and `@RangeDouble` would be as follows.
+
+````Java
```
Comment (elifoster) [inline docs/utilities/config_annotations.md]: This should only be 3 backticks

File: docs/utilities/config_annotations.md [hunk 26]
```diff
@@ -0,0 +1,234 @@
...
+  @Comment("This integer property accepts values 1 to 10 inclusive.")
+  @RangeInt(min = 1, max = 10)
+  public static int myIntegerProperty = false;
+
+  @Comment("This double property accepts values 0.0 to 1.0 inclusive.")
+  @RangeDouble(min = 0.0, max = 1.0)
+  public static double myDoubleProperty = false;
+
+}
+````
```
Comment (elifoster) [inline docs/utilities/config_annotations.md]: This should only be 3 backticks

File: docs/utilities/config_annotations.md [hunk 27]
```diff
@@ -0,0 +1,234 @@
...
+
+!! important
+
+    While both the `@RangeInt` and `@RangeDouble` accept a min and a max value to be set they are not interchangeable, given that the type must match. Consequently the `@RangeInt` annotation is only to be applied to `int` and `Integer` fields, and the `@RangeDouble` is only to be applied to `double` and `Double` fields.
+
+!! important
```
Comment (elifoster) [inline docs/utilities/config_annotations.md]: Having 2 `!! important`s in a row might be weird.

File: docs/utilities/config_annotations.md (hunk 7 above)
Comment (elifoster) [inline docs/utilities/config_annotations.md]: I think all of the backticked types are unnecessary. I'd rework this to be

>There currently exists no range annotations capable of supporting numeric types other than the aforementioned integer and double types. As such, you are forced to use integers instead of bytes and longs, and doubles instead of floating point real numbers.

or something.

File: docs/utilities/config_annotations.md [hunk 28]
```diff
@@ -0,0 +1,234 @@
...
+
+    There currently exists no range annotations capable of supporting numeric types other than the aforementioned `int`, `Integer`, `double`, and `Double` types. As such, you are forced to use `int` instead of `byte` or `long`, `Integer` instead of `Byte` or `Long`, `double` instead of `float`, and `Double` instead of `Float`. This issue will likely be fixed in a future release given wide enough adoption of the new annotation-based configuration system.
+
+
+Sub-categories
+-------------------------
+
+Smaller mods will generally have fewer properties (say about 5 - 10) and do not need to worry about dividing up their configurations into more manageable groups of properties for the end-user to manage. However, for larger mods configurations consisting of 20 to 50 elements is not all that uncommon, and the need for sub-categories becomes irrefutable. Recognizing such a need, the annotation-based configuration system provides a mechanism by which `@Config` classes can be divided into groups that are reflected directly in the structure of the `@Config` class.
```
Comment (elifoster) [inline docs/utilities/config_annotations.md]: The comparison of small/large mods is totally unnecessary here. People should be able to analyze whether they need subcategories on their own based on their own needs.

File: docs/utilities/config_annotations.md [hunk 29]
```diff
@@ -0,0 +1,234 @@
...
+
+The sub-division of `@Config` classes is accomplished by the addition of `public static` inner classes that act as the physical configuration groupings. Thus, any property that is placed in the inner class is then resolved by the system as not to exist in the default group defined by the `@Config` annotation but rather a new group with the name of the inner class.
```
Comment (elifoster) [inline docs/utilities/config_annotations.md]: Do these classes actually have to be public?

File: docs/utilities/config_annotations.md [hunk 30]
```diff
@@ -0,0 +1,234 @@
...
+
+Building upon the prior examples, an example usage of property grouping is as follows:
+
+````Java
```
Comment (elifoster) [inline docs/utilities/config_annotations.md]: This should be 3 backticks

File: docs/utilities/config_annotations.md [hunk 31]
```diff
@@ -0,0 +1,234 @@
...
+
+  public static class MyCategory {
+
+    @Comment("This property belongs to the MyCategory group.")
+    public static boolean myPropertyMyCategory = false;
+
+  }
+
+}
+````
```
Comment (elifoster) [inline docs/utilities/config_annotations.md]: This should be 3 backticks

File: docs/utilities/config_annotations.md [hunk 32]
```diff
@@ -0,0 +1,234 @@
...
+
+
+Alternate Names
+-------------------------
+
+Sometimes, for reasons unknown, we wind up with variable names in a configuration class that would be unsuitable for use as an actual property key. This would be problematic as Forge determines the property key based off of variable names. For this reason, the `@Name` annotation exists.
```
Comment (elifoster) [inline docs/utilities/config_annotations.md]: Too much fluff. I'd reword this to

>Using the `@Name` annotation, the property can have a key or name specified that is different from its field's name.

Really that's all the information that's needed here.

File: docs/utilities/config_annotations.md [hunk 33]
```diff
@@ -0,0 +1,234 @@
...
+
+The name annotation allows for the convenient changing of the property's name from the default value inferred from the name of the variable to something that is deemed more suitable by the modder.
```
Comment (elifoster) [inline docs/utilities/config_annotations.md]: Redundant

File: docs/utilities/config_annotations.md [hunk 34]
```diff
@@ -0,0 +1,234 @@
...
+
+An example usage would be as follows:
+
+````Java
```
Comment (elifoster) [inline docs/utilities/config_annotations.md]: This example is not necessary.

File: docs/utilities/config_annotations.md (hunk 8 above)
Comment (elifoster) [inline docs/utilities/config_annotations.md]: "conventional" is subjective. My conventions are likely different from your conventions, which are likely different from tterrag's conventions, which are likely different from the conventions specified by Oracle.

File: docs/utilities/config_annotations.md [hunk 35]
```diff
@@ -0,0 +1,234 @@
...
+  public static boolean MY_hoRibbliyNAMEdVairable = false;
+
+}
+````
+
+!! Note
+
+    If you are following conventional Java variable naming practices then the variables defined in your `@Config` class likely contain a great number of capital letters, as seen in the provided example usages (prior to the horribly named one). Since the Forge standard for registration names is all-lowercase, the default property names would therefore be unsuitable, and use of the `@Name` annotation is highly recommended.
+
+    This behavior is partly an oversight and partly a technical limitation. Forge has no intention of punishing any modder who conforms to standard practices. Ideally future releases of Forge will automatically convert traditionally-named variables into conforming property names by replacing capital characters with respective lowercase characters preceded by the underscore character `_`.
```
Comment (elifoster) [inline docs/utilities/config_annotations.md]: >Forge has no intention of punishing any modder who conforms to standard practices.

Why would they?

Comment (elifoster) [inline docs/utilities/config_annotations.md]: This is unnecessary speculation.

File: docs/utilities/config_annotations.md [hunk 36]
```diff
@@ -0,0 +1,234 @@
...
+
+!! Note
```
Comment (elifoster) [inline docs/utilities/config_annotations.md]: This being in a note is not necessary imo.

File: docs/utilities/config_annotations.md (hunk 9 above)
Comment (elifoster) [inline docs/utilities/config_annotations.md]: >to a inner

>to an inner

File: docs/utilities/config_annotations.md [hunk 37]
```diff
@@ -0,0 +1,234 @@
...
+  public static class MyCategory {
+
+      /* ... */
+
+  }
+
+}
+````
+
+!! Important
```
Comment (elifoster) [inline docs/utilities/config_annotations.md]: `!! important` lowercase i

File: docs/utilities/config_annotations.md [hunk 38]
```diff
@@ -0,0 +1,234 @@
...
+
+    Applying this annotation to a category does not alter the language key of properties contained in that category.
+
+
+Time Specific Properties
```
Comment (elifoster) [inline docs/utilities/config_annotations.md]: Time-Specific

File: docs/utilities/config_annotations.md [hunk 39]
```diff
@@ -0,0 +1,234 @@
...
+-------------------------
+
+Occasionally in order for a property change to take effect, a mod may require either a complete restart of Minecraft/Forge or a restart of the world. This distinction is important, as a property may control whether or not a sub-module is loaded or not, and a sudden change in the value of the property might cause the system to go looking for an unloaded sub-module. Ideally, the sub-module could be loaded on-the-fly, but in the case of a sub-module that adds a block to the game, attempting to load the sub-module would cause an error as blocks can only be added during initialization phases. Consequently, there exists a need as to signal the configuration system as to hold off making a change to the value specific properties in a `@Config` class until either a world reload occurs or a game reload occurs. For this reason, the annotations `@RequiresWorldRestart` and `@RequiresMcRestart` exist.
```
Comment (elifoster) [inline docs/utilities/config_annotations.md]: Too much fluff.

>Occasionally in order for a property change to take effect, a mod may require either a complete restart of Minecraft/Forge or a restart of the world. The `@RequiresWorldRestart` and `RequiresMcRestart` annotations prevent the annotated value from being modified before a world or game restart respectively.

File: docs/utilities/config_annotations.md [hunk 40]
```diff
@@ -0,0 +1,234 @@
...
+
+The `@RequiresWorldRestart` annotation works by holding off applied changes to a configuration file until a world reload occurs, at which point the changes are then applied, and the world reloaded.
```
Comment (elifoster) [inline docs/utilities/config_annotations.md]: Already explained

Comment (elifoster) [inline docs/utilities/config_annotations.md]: Does the world reload before or after the change is made? By this sentence, the sequence would be: restart, change value, restart again. I highly doubt that's what is actually done.

File: docs/utilities/config_annotations.md [hunk 41]
```diff
@@ -0,0 +1,234 @@
...
+  public static boolean myWorldReloadChangeableProperty = false;
+
+  @Comment("This property is only changeable on Minecraft restarts!")
+  @RequiresMcRestart
+  public static boolean myMcReloadChangeableProperty = false;
+
+}
+````
+
+!! Important
```
Comment (elifoster) [inline docs/utilities/config_annotations.md]: `!! important` lowercase i

File: docs/utilities/config_annotations.md [hunk 42]
```diff
@@ -0,0 +1,234 @@
...
+
+    Keep in mind that for installs including more than just one or two other mods can take a great deal of time to restart. As such you should try to make as many properties as possible changeable on the fly.
+
+!! Important
```
Comment (elifoster) [inline docs/utilities/config_annotations.md]: `!! important` lowercase i

File: docs/utilities/config_annotations.md [hunk 43]
```diff
@@ -0,0 +1,234 @@
...
+  @Comment("This property is only changeable on Minecraft restarts!")
+  @RequiresMcRestart
+  public static boolean myMcReloadChangeableProperty = false;
+
+}
+````
+
+!! Important
+
+    Keep in mind that for installs including more than just one or two other mods can take a great deal of time to restart. As such you should try to make as many properties as possible changeable on the fly.
```
Comment (elifoster) [inline docs/utilities/config_annotations.md]: Unnecessary comment.

File: docs/utilities/config_annotations.md [hunk 44]
```diff
@@ -0,0 +1,234 @@
...
+-------------------------
+
+| Field         | Manditory     | Description
+| ------------- | ------------- | -----------
+| `modid`       | Yes           | The id of the mod that this configuration is associated with.
+| `name`        | No            | The user-friendly name of the configuration file that this class represents. If this field is not set manually, it will default to the value set in the `modid` field.
+| `type`        | No            | The 'type' of this configuration object, currently the only valid value is the default, `Type.INSTANCE`. More configuration types are to be expected in future releases.
+| `category`    | No            | The 'root' category for all elements contained in this configuration object. If the field is not set manually, then it will default to `"general"`. Conversely, if the field is set to `""` the root category will be disabled, and all internal elements will need a category to be manually set with a different annotation (not recommended).
+
+!! Important
```
Comment (elifoster) [inline docs/utilities/config_annotations.md]: `!! important` lowercase i

File: docs/utilities/config_annotations.md (hunk 29 above)
Comment (TehNut) [inline docs/utilities/config_annotations.md]: They *do not* have to be inner classes.

File: docs/utilities/config_annotations.md (hunk 1 above)
Comment (RlonRyan) [reply docs/utilities/config_annotations.md]: An easily missed technicality of the system that should indeed be clearly documented.

File: docs/utilities/config_annotations.md (hunk 2 above)
Comment (RlonRyan) [reply docs/utilities/config_annotations.md]: ... fields in a '@Config' class.

File: docs/utilities/config_annotations.md (hunk 5 above)
Comment (RlonRyan) [reply docs/utilities/config_annotations.md]: Understood.

File: docs/utilities/config_annotations.md (hunk 6 above)
Comment (RlonRyan) [reply docs/utilities/config_annotations.md]: The goal was consistency within each of the sections, consisting of a description followed by a code example finalized by notes (i.e. important, note).

File: docs/utilities/config_annotations.md (hunk 7 above)
Comment (RlonRyan) [reply docs/utilities/config_annotations.md]: This was from the direct analysis of the code. The `@RangeInt` is typed to integer values so it cannot cover the actual range of long values. In the case of short and byte, `@RangeInt` will accept values that are greater than the respective maximum values of the types, causing silent overflow. While the code might not crash, it may lead to the aggravation of end-users wondering as to why a field keeps getting set to a seemingly arbitrary number and not the high-value number they specified.

In the case of `@RangeDouble`, the same overflow/underflow issue would be present, however, the behavior would be even less well defined, given the significant difference in precision of the two formats.

The intention was not to speculate here, but rather inform the user that they should keep an eye out for new annotations possibly introduced in future versions, and write their code in a future-proof way.

File: docs/utilities/config_annotations.md (hunk 9 above)
Comment (RlonRyan) [reply docs/utilities/config_annotations.md]: Again, the intention was not to speculate baselessly, but rather to inform the user that the system is relatively new, and to keep an eye out for changes that might impact the way that they code or influence their adoption of the system.

File: docs/utilities/config_annotations.md (hunk 13 above)
Comment (RlonRyan) [reply docs/utilities/config_annotations.md]: The system will skip any non-public fields.

File: docs/utilities/config_annotations.md (hunk 3 above)
Comment (RlonRyan) [reply docs/utilities/config_annotations.md]: My bad.

File: docs/utilities/config_annotations.md (hunk 18 above)
Comment (RlonRyan) [reply docs/utilities/config_annotations.md]: The goal was to stress the motivation behind the `@Comment` annotation, given that understanding the reasoning behind a given element can provide great insight into the proper usage of the element.

File: docs/utilities/config_annotations.md (hunk 7 above)
Comment (TehNut) [reply docs/utilities/config_annotations.md]: But we do not know if these annotations will ever be added, so it is speculation.

File: docs/utilities/config_annotations.md (hunk 8 above)
Comment (RlonRyan) [reply docs/utilities/config_annotations.md]: Java convention is well-defined as per [The Java Language Specification](https://docs.oracle.com/javase/specs/jls/se8/html/index.html). See Chapter 6 for an in-depth description of the Java naming system and associated convention.

File: docs/utilities/config_annotations.md (hunk 29 above)
Comment (RlonRyan) [reply docs/utilities/config_annotations.md]: That is a good point that should be added to the documentation.

File: docs/utilities/config_annotations.md (hunk 13 above)
Comment (elifoster) [reply docs/utilities/config_annotations.md]: This should be clearly documented then.

File: docs/utilities/config_annotations.md (hunk 8 above)
Comment (elifoster) [reply docs/utilities/config_annotations.md]: The Java "official" convention is still subjective and many people don't use it because they don't like it or haven't read it. Personally I've never even read through it in its entirety and I've been using Java for about half a decade now (I based my code style off of Google's standards roughly, if I recall correctly).

File: docs/utilities/config_annotations.md (hunk 35 above)
Comment (LexManos) [reply docs/utilities/config_annotations.md]: >Ideally future releases of Forge will automatically

Never gunna happen, your shit is your shit. Stop talking about standards/speculating on Forge and just talk about the functionality.

File: docs/utilities/config_annotations.md (hunk 34 above)
Comment (Akxe) [reply docs/utilities/config_annotations.md]: I as someone new to this love the possibility to just copy and paste example and fiddle with it in my code. Just as a note... You can always scroll, it ain't hard.

**Comment by Akxe** [comment]
I have found most of things I wanted to know in here, or was able to figure out the rest base on these information, but one. How do one translate comments. The `@Comment` annotation accepts a string, that will be shown to player and also will be copied to `modid.cfg` file.

**Comment by Akxe** [comment]
@Mod's `canBeDeactivated` setting should also be explained here, as it is almost part of mod configuration.

**Comment by Akxe** [comment]
I was also unable to find any events, that would indicate, that one of those properties got changed.

File: docs/utilities/config_annotations.md (hunk 34 above)
Comment (LexManos) [reply docs/utilities/config_annotations.md]: Yes but i've mandated that we are not a copy/paste documentation. That leads to nothing but people being lazy and not learning the actual subject. And before you argue yes i've heard it all before. However my direction sill stands. No functional copy pasta allowed in our docs.

Comment (elifoster) [reply docs/utilities/config_annotations.md]: @LexManos As sort of a side thing it might be useful to put that mandate into a `CONTRIBUTING.md` file or something similar.

**Comment by tterrag1098** [comment]
Superceded by #147
